holmes/core/tool_calling_llm.py

```diff
@@ -144,30 +146,40 @@ def call(self, system_prompt, user_prompt, post_process_prompt: Optional[str] =
             if response_message.content and ('bedrock' not in self.model and logging.DEBUG != logging.root.level):
                 logging.warning(f"got unexpected response when tools were given: {response_message.content}")
 
-            for t in tools_to_call:
-                tool_name = t.function.name
-                tool_params = json.loads(t.function.arguments)
-                tool = self.tool_executor.get_tool_by_name(tool_name)
-                tool_response = tool.invoke(tool_params)
-                MAX_CHARS = 100_000 # an arbitrary limit - we will do something smarter in the future
-                if len(tool_response) > MAX_CHARS:
-                    logging.warning(f"tool {tool_name} returned a very long response ({len(tool_response)} chars) - truncating to last 10000 chars")
-                    tool_response = tool_response[-MAX_CHARS:]
-                messages.append(
-                    {
-                        "tool_call_id": t.id,
-                        "role": "tool",
-                        "name": tool_name,
-                        "content": tool_response,
-                    }
-                )
-                tool_calls.append(
-                    ToolCallResult(
-                        tool_name=tool_name,
-                        description=tool.get_parameterized_one_liner(tool_params),
-                        result=tool_response,
+            with concurrent.futures.ThreadPoolExecutor(max_workers=16) as executor:
+                futures = []
+                for t in tools_to_call:
+                    futures.append(executor.submit(self._invoke_tool, t))
+
+                for future in concurrent.futures.as_completed(futures):
+                    tool_call_result: ToolCallResult = future.result()
+                    tool_calls.append(tool_call_result)
+                    messages.append(
+                        {
+                            "tool_call_id": tool_call_result.tool_call_id,
+                            "role": "tool",
+                            "name": tool_call_result.tool_name,
+                            "content": tool_call_result.result,
+                        }
                     )
-                )
+
+    def _invoke_tool(self, tool_to_call):
```

Comment:
Can you add the return type?
Also, what is the type of `tool_to_call`? What fields are in there?

Reply:
Fixed